docs/examples/howto/add_new_algorithm.rst

```diff
@@ -0,0 +1,180 @@
+An algorithm is stateless if does not modify its own attributes while executing.
+This way it becomes reentrant and in consequence thread safe.
+For instance if the event processing is best organized with such methods:
+
+.. code-block:: cpp
+
+    void prepareBuffers();
+    void fillBuffers( const SpacePoint& );
+    void extractInfo();
+
+that need to pass the data between each other these methods should rather look like this:
+
+.. code-block:: cpp
+
+    struct SeedingBuffers{ ... some buffers ... };
+    void prepareBuffers(SeedingBuffers& buffers);
+    void fillBuffers(SeedingBuffers& buffers, const SpacePoint& );
+    void extractInfo(const SeedingBuffers& buffers);
+
+
+..  tip:: It is common practice to put the algorithm code in the `ActsExamples` namespace.
+
+Input and Output
+------------------
+
+The algorithm would be typically part of some processing sequence 
+and thus consume and produce some event data. 
+In the hypothetical example discussed here, 
+space-points are the input and track seeds an output. 
+The data can be retrieved in the algorithm using the `get` method:
+
+.. code-block:: cpp
+
+    template<typename T>
+    const T& get(const std::string& name);
+
+The data is fetched from the "store" that is populated by a preceding algorithm or by the reader.
+
+The data object (or objects) produced by an algorithm can be placed in the store using an `add` method:
+
+.. code-block:: cpp
+
+    template<typename T>
+    void add(const std::string& name, T&& object);
+
+The ownership of the object is transferred to the store. 
+That is, the destruction of this object at the end of event processing is taken care of.
+
+Configurability
+----------------
+
+It is customary that an algorithm requires configuration parameters. 
+For example, in a seeding algorithm these parameters could include which detector layers should be used. 
+The configuration can be provided to an algorithm through an additional class/structure. 
+It can be an inner class of the algorithm or it can be external to it. 
+One should use an external structure if it is shared among several algorithm classes.
+
+For example that is how the configuration object could look like for `MySeedingAlgorithm`:
+
+.. code-block:: cpp
+
+    struct MySeedingConfig {
+        std::vector<int> layers; // layers to use by the seeder
+        float deltaZ;  // the maximum allowed deviation in r-z plane
+    };
+
+.. tip:: It is customary to put the config structures in the ``Acts`` namespace.
+
+The algorithm constructor would take a `MySeedingConfig` object during 
+the construction in addition to an argument controlling verbosity of diagnostic messages.
+
+.. code-block:: cpp
+
+    MySeedingAlgorithm::MySeedingAlgorithm( Acts::MySeedingConfig cfg, Acts::Logging::Level lvl):
+      ActsExamples::BareAlgortihm("MySeedingAlgorithm", lvl), 
+      m_cfg(std::move(cfg)){...}
+
+
+Python bindings
```

Comment:
I guess, we'll update this if #1510 is merged.

Reply:
Yes, exactely. This is I think the most common design we have. 
So this macro is also not designed to be ultra flexible and customizable, but to cover the most basic case. There are several algorithms that are not implemented with this, because they require e.g, additional member functions or something else.
But I would think this is okay for a "my-first-algorithm" guide, but its definitively good to point out the assumptions behind it and maybe also say that its not that difficult to implement more complicated things.

What I think is a bit misleading now is, that you have one section of defining the config struct, and one of the algorithm, where you show both the macro and the "manual" binding. But the macro also includes defining the binding to the config struct, so it is not necessary to define this seperatly using `ACTS_PYTHON_STRUCT_BEGIN` etc. 

So maybe it would be good to swap this, so you have one code block with the macro, and another with the "manual binding"?

But thanks for adapting this anyways :)

Reply:
Actually, I think we typically have this situation. I.e. the Alg::Config is an internal class. The Core algorithm codes config is something separate. let me redo the instruction in this spirit.

Reply:
The update is pushed.

Reply:
@benjaminhuth - I would like to resolve this thread as I addressed your comments. Let me know.

Reply:
Resolving

Reply:
Ah I somehow missed that this was still open. Sorry for the delay!